src/pushed_authorization.rs

```diff
@@ -0,0 +1,293 @@
+                http_response.body,
+                "unexpected HTTP status code".to_string(),
+            ));
+        }
+
+        let parsed_response: PushedAuthorizationResponse = match http_response
+            .headers
+            .get(CONTENT_TYPE)
+            .map(ToOwned::to_owned)
+            .unwrap_or_else(|| HeaderValue::from_static(MIME_TYPE_JSON))
+        {
+            ref content_type if content_type_has_essence(content_type, MIME_TYPE_JSON) => {
+                serde_path_to_error::deserialize(&mut serde_json::Deserializer::from_slice(
+                    &http_response.body,
+                ))
+                .map_err(RequestError::Parse)
+            }
+            ref content_type => Err(RequestError::Response(
+                http_response.status_code,
+                http_response.body,
+                format!("unexpected response Content-Type: `{:?}`", content_type),
+            )),
+        }?;
+
+        auth_url
+            .query_pairs_mut()
+            .append_pair("request_uri", parsed_response.request_uri.get());
+
+        auth_url
+            .query_pairs_mut()
+            .append_pair("client_id", &req_body.client_id.to_string());
+
+        Ok((auth_url, token))
+    }
+
+    pub fn prepare_request<RE>(
+        self,
+        client_assertion: Option<String>,
+        client_assertion_type: Option<String>,
+    ) -> Result<(HttpRequest, ParAuthParams, CsrfToken), RequestError<RE>>
+    where
+        RE: std::error::Error + 'static,
+    {
+        let (url, token) = self.inner.url();
+
+        let body = serde_urlencoded::from_str::<ParAuthParams>(url.clone().as_str())
+            .map_err(|_| RequestError::Other("failed parsing url".to_string()))?
+            .set_client_assertion_type(client_assertion_type.clone())
+            .set_client_assertion(client_assertion.clone())
+            .set_authorization_details(Some(
+                serde_json::to_string::<Vec<AuthorizationDetail<AD>>>(&self.authorization_details)
+                    .unwrap_or("[]".to_string()),
+            ))
+            .set_wallet_issuer(self.wallet_issuer)
+            .set_user_hint(self.user_hint)
+            .set_issuer_state(self.issuer_state);
+
+        Ok((
+            HttpRequest {
+                url: self.par_auth_url.url().clone(),
+                method: Method::POST,
+                headers: vec![
+                    (
+                        CONTENT_TYPE,
+                        HeaderValue::from_static(MIME_TYPE_FORM_URLENCODED),
+                    ),
+                    (ACCEPT, HeaderValue::from_static(MIME_TYPE_JSON)),
+                ]
+                .into_iter()
+                .collect(),
+                body: serde_json::to_vec(&format!(
+                    "&{:}&", // had to add & before and after to fix parsing during the request
+                    serde_urlencoded::to_string(&body).unwrap_or("".to_string())
+                ))
```

Comment:
Perhaps I'm wrong, but I believe you can do this:

```suggestion
                body: serde_urlencoded::to_string(&body).map_err(...)?.as_bytes().to_vec(),
```